esp/logging/loggingagent/cassandraloggingagent/cassandralogagent.cpp

```diff
@@ -0,0 +1,621 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2016 HPCC Systems.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "LoggingErrors.hpp"
+#include "esploggingservice_esp.ipp"
+#include "cassandralogagent.hpp"
+
+using namespace cassandraembed;
+
+const int DefaultMaxTriesGTS = -1;
+const char* const DefaultloggingTransactionApp = "CassandraloggingTransaction";
+const char* const DefaultTransactionApp = "DefauleTransaction";
+
+static void setCassandraLogAgentOption(StringArray& opts, const char* opt, const char* val)
+{
+    if (opt && *opt && val)
+    {
+        VStringBuffer optstr("%s=%s", opt, val);
+        opts.append(optstr);
+    }
+}
+
+static const CassValue* getSingleResult(const CassResult* result)
+{
+    const CassRow* row = cass_result_first_row(result);
+    return row ? cass_row_get_column(row, 0) : NULL;
+}
+
+void ensureInputString(const char* input, bool lowerCase, StringBuffer& inputStr, int code, const char* msg)
+{
+    inputStr.set(input).trim();
+    if (inputStr.isEmpty())
+        throw MakeStringException(code, "%s", msg);
+    if (lowerCase)
+        inputStr.toLowerCase();
+}
+
+void CLogTable::loadMappings(IPropertyTree& fieldList)
+{
+    Owned<IPropertyTreeIterator> itr = fieldList.getElements("Field");
+    ForEach(*itr)
+    {
+        IPropertyTree &map = itr->query();
+
+        StringBuffer name, mapTo, fieldType, defaultValue;
+        ensureInputString(map.queryProp("@name"), false, name, -1, "Field @name required");
+        ensureInputString(map.queryProp("@mapto"), true, mapTo, -1, "Field @mapto required");
+        ensureInputString(map.queryProp("@type"), true, fieldType, -1, "Field @type required");
+        defaultValue = map.queryProp("@default");
+        defaultValue.trim();
+
+        Owned<CLogField> field = new CLogField(name.str(), mapTo.str(), fieldType.str());
+        if (!defaultValue.isEmpty())
+            field->setDefault(defaultValue.str());
+        logFields.append(*field.getClear());
+    }
+}
+
+void CLogGroup::loadMappings(IPropertyTree& fieldList)
+{
+    Owned<IPropertyTreeIterator> itr = fieldList.getElements("Fieldmap");
+    ForEach(*itr)
+    {
+        StringBuffer tableName;
```

Comment:
The constructing and destruction of the StringBuffer is expensive and should be avoid if possible, I believe the content of the stringbuffer, the tablename, is copied in the CLogTable making this new stringbuffer instance per iteration wasteful.